api/src/main/java/org/apache/iceberg/stats/StatsUtil.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.stats;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.types.Types;
+
+public class StatsUtil {
+  static final int COLUMN_SIZE_OFFSET = 0;
+  static final int VALUE_COUNT_OFFSET = 1;
+  static final int NULL_VALUE_COUNT_OFFSET = 2;
+  static final int NAN_VALUE_COUNT_OFFSET = 3;
+  static final int LOWER_BOUND_OFFSET = 4;
+  static final int UPPER_BOUND_OFFSET = 5;
+  static final int NUM_STATS_PER_COLUMN = 200;
+  static final int RESERVED_FIELD_IDS = 200;
+  static final int DATA_SPACE_FIELD_ID_START = 10_000;
+  static final int METADATA_SPACE_FIELD_ID_START = 2_147_000_000;
+  static final int RESERVED_FIELD_IDS_START = Integer.MAX_VALUE - RESERVED_FIELD_IDS;
+  // support stats for only up to this amount of data field IDs
+  static final int MAX_DATA_FIELD_ID = 1_000_000;
+
+  private StatsUtil() {}
+
+  public static int statsFieldIdForField(int fieldId) {
+    return fieldId >= RESERVED_FIELD_IDS_START
+        ? statsFieldIdForReservedField(fieldId)
+        : statsFieldIdForDataField(fieldId);
+  }
+
+  private static int statsFieldIdForDataField(int fieldId) {
+    long statsFieldId = DATA_SPACE_FIELD_ID_START + NUM_STATS_PER_COLUMN * (long) fieldId;
+    if (fieldId > MAX_DATA_FIELD_ID || statsFieldId < 0) {
+      return -1;
+    }
+
+    return (int) statsFieldId;
+  }
+
+  private static int statsFieldIdForReservedField(int fieldId) {
+    int offset = RESERVED_FIELD_IDS - (Integer.MAX_VALUE - fieldId);
+
+    long statsFieldId = METADATA_SPACE_FIELD_ID_START + NUM_STATS_PER_COLUMN * (long) offset;
+    if (statsFieldId < 0 || statsFieldId > RESERVED_FIELD_IDS_START) {
+      // ID overflows
```

Comment:
Shall we at least log a warn?
AFAIK, we didn't plan for this

Reply:
I think the only concern I had with the warn is if a user had a large schema with lots and lots of "invalid" (e.g. > 1_000_000) field IDs but practically speaking I don't think this will ever be a concern in a real use case. I've added a WARN stating that the stats schema for those field IDs couldn't be created